### PR TITLE
brie: add error info column and history backup/restore info in sql

### PR DESCRIPTION
**Pull request by lichunzhu**
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

Issue Number: close #xxx <!-- REMOVE this line if no issue to close -->

Problem Summary: When we use BRIE to run a backup job, if our mysql connection is broken due to some reason, we can't infer whether this BRIE job is successful or not.

### What is changed and how it works?

What's Changed: Save history brie jobs in the task queue and add an error info column.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
Run a BR job and stop the network.
Current:
```
mysql> show backups;
+----------------------+--------+-------------------+---------------------+---------------------+---------------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Destination          | State  | Progress          | Queue_time          | Execution_time      | Finish_time         | Connection | Message                                                                                                                                                                                                                                         |
+----------------------+--------+-------------------+---------------------+---------------------+---------------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| s3://chunzhu/backup/ | Backup | 83.33333333333333 | 2021-02-03 17:46:52 | 2021-02-03 17:46:52 | 2021-02-03 17:48:31 |          3 | [executor:8124]Backup failed: msg:"Io(Custom { kind: Other, error: \"failed to put object Error during dispatch: error trying to connect: tcp connect error: Connection refused (os error 111)\" })" : [BR:KV:ErrKVUnknown]unknown tikv error |
+----------------------+--------+-------------------+---------------------+---------------------+---------------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
Previous:
No backup can be found.

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note <!-- bugfixes or new feature need a release note -->

- No release note

### CODE REVIEW

File: executor/brie.go [hunk 1]
```diff
@@ -377,11 +407,17 @@ func (e *ShowExec) fetchShowBRIE(kind ast.BRIEKind) error {
...
 			e.result.AppendTime(3, item.info.queueTime)
 			e.result.AppendTime(4, item.info.execTime)
-			e.result.AppendNull(5) // FIXME: fill in finish time after keeping history.
+			e.result.AppendTime(5, item.info.finishTime)
 			e.result.AppendUint64(6, item.info.connID)
+			if len(item.info.errStr) > 0 {
+				e.result.AppendString(7, item.info.errStr)
+			} else {
+				e.result.AppendNull(7)
+			}
```
Comment (overvenus) [inline executor/brie.go:416]: Could you add a test to cover them?

**Comment by kennytm** [comment]
urgh, but legit.

```
[2021-02-26T11:25:59.060Z] ----------------------------------------------------------------------
[2021-02-26T11:25:59.060Z] FAIL: brie_test.go:61: testBRIESuite.TestFetchShowBRIE
[2021-02-26T11:25:59.060Z] 
[2021-02-26T11:25:59.060Z] brie_test.go:122:
[2021-02-26T11:25:59.060Z]     c.Assert(res, Equals, info1Res)
[2021-02-26T11:25:59.060Z] ... obtained string = "noop://test/backup1, Wait, 0, 2021-02-26 18:51:53, 2021-02-26 18:51:53, 2021-02-26 18:51:53, 0, killed\n"
[2021-02-26T11:25:59.060Z] ... expected string = "noop://test/backup1, Wait, 0, 2021-02-26 18:51:53, 2021-02-26 18:51:53, 2021-02-26 18:51:53, 0, killed"
[2021-02-26T11:25:59.060Z] 
[2021-02-26T11:25:59.060Z] PASS: split_test.go:369: testSplitIndex.TestClusterIndexSplitTable	0.014s
[2021-02-26T11:25:59.060Z] PASS: split_test.go:48: testSplitIndex.TestLongestCommonPrefixLen	0.000s
```

**Comment by overvenus** [comment]
run-integration-br-test

File: executor/brie.go [hunk 2]
```diff
@@ -46,6 +46,13 @@ import (
 	"github.com/pingcap/tidb/util/sqlexec"
 )
 
+const clearInterval = 10 * time.Minute
+
+var outdatedDuration = types.Duration{
```
Comment (overvenus) [inline executor/brie.go:51]: Can it be a const?
```suggestion
const outdatedDuration = types.Duration{
```

File: util/chunk/row.go [hunk 3]
```diff
@@ -221,35 +221,35 @@ func (r Row) ToString(ft []*types.FieldType) string {
...
-			buf = append(buf, r.GetJSON(colIdx).String()...)
-		case types.ETReal:
-			switch ft[colIdx].Tp {
-			case mysql.TypeFloat:
-				buf = strconv.AppendFloat(buf, float64(r.GetFloat32(colIdx)), 'f', -1, 32)
-			case mysql.TypeDouble:
-				buf = strconv.AppendFloat(buf, r.GetFloat64(colIdx), 'f', -1, 64)
+			buf = append(buf, "NULL"...)
+		} else {
+			switch ft[colIdx].EvalType() {
```
Comment (overvenus) [inline util/chunk/row.go:226]: Why replace `continue` with `else`?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: executor/brie.go (hunk 2 above)
Comment (lichunzhu) [reply executor/brie.go:51]: No. Go doesn't support struct constants. https://golang.org/ref/spec#Constants

File: util/chunk/row.go (hunk 3 above)
Comment (lichunzhu) [reply util/chunk/row.go:226]: If `colIdx` is the last column of this row, we shouldn't put the comma(`, `) after `nil` nor `NULL`.

**Review by overvenus — COMMENTED** [review]
LGTM

**Review by winoros — APPROVED** [review]

**Review by kennytm — APPROVED** [review]

**Comment by kennytm** [comment]
/run-check_dev_2

https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/tidb_ghpr_check_2/runs/17857/nodes/64/log/?start=0

stuck on TestCoprocessorOOMAction

**Pull request merged**